tracer/src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.NetCore.cs

```diff
@@ -46,26 +61,79 @@ private static Assembly AssemblyResolve_ManagedProfilerDependencies(object sende
             {
+                StartupLogger.Debug("Assembly {0} not found.", args.Name);
                 return null;
             }
 
             var path = Path.Combine(ManagedProfilerDirectory, $"{assemblyName.Name}.dll");
 
-            // Only load the main profiler into the default Assembly Load Context.
-            // If Datadog.Trace or other libraries are provided by the NuGet package their loads are handled in the following two ways.
-            // 1) The AssemblyVersion is greater than or equal to the version used by Datadog.Trace, the assembly
-            //    will load successfully and will not invoke this resolve event.
-            // 2) The AssemblyVersion is lower than the version used by Datadog.Trace, the assembly will fail to load
-            //    and invoke this resolve event. It must be loaded in a separate AssemblyLoadContext since the application will only
-            //    load the originally referenced version
-            if (File.Exists(path))
+            if (TryGetCachedAssembly(path, out var cachedAssembly))
             {
+                // The file exists in the Home folder...
+                if (cachedAssembly is not null)
+                {
+                    // The assembly is already loaded.
+                    return cachedAssembly;
+                }
+
+                // Only load the main profiler into the default Assembly Load Context.
+                // If Datadog.Trace or other libraries are provided by the NuGet package their loads are handled in the following two ways.
+                // 1) The AssemblyVersion is greater than or equal to the version used by Datadog.Trace, the assembly
+                //    will load successfully and will not invoke this resolve event.
+                // 2) The AssemblyVersion is lower than the version used by Datadog.Trace, the assembly will fail to load
+                //    and invoke this resolve event. It must be loaded in a separate AssemblyLoadContext since the application will only
+                //    load the originally referenced version
                 StartupLogger.Debug("Loading {0} with DependencyLoadContext.LoadFromAssemblyPath", path);
-                return DependencyLoadContext.LoadFromAssemblyPath(path); // Load unresolved framework and third-party dependencies into a custom Assembly Load Context
+                var assembly = DependencyLoadContext.LoadFromAssemblyPath(path); // Load unresolved framework and third-party dependencies into a custom Assembly Load Context
+                TrySetCachedAssembly(path, assembly);
+                return assembly;
             }
 
+            // The file doesn't exist in the Home folder.
             return null;
         }
+
+        private static bool TryGetCachedAssembly(string path, out Assembly cachedAssembly)
+        {
+            for (var i = 0; i < _assemblies.Length; i++)
```

Comment:
I think you need a `var assemblies = _assemblies` here, as there's no locking around `_assemblies`? Alternatively (probably preferably), could add an `Interlocked.CompareExchange` in `ResolveManagedProfilerDirectory` so we don't replace it if it's already set

Reply:
OK, so `_assemblies` will only be set once _assuming_ ResolveManagedProfilerDirectory is only called once (it currently is)